File: hh-credit-autoload-ruby/services/lambda_function.js
const HelloHunterApi = require('../functions/hello-hunter-api');
const ZohoCrmApi = require('../functions/zoho-crm-api');
const logger = require('winston');

class HelloHunterService {
    constructor() {
        this.hhApi = new HelloHunterApi();
        this.crmApi = new ZohoCrmApi();
    }

    async initHhClient() {
        const hhCustomers = await this.hhApi.getCustomers();
        logger.info('Initialized Hello Hunter Client');
        return hhCustomers;
    }

    async initZClient() {
        const query = "select Account_Name, Hello_Hunter_Id, Threshold, Reload_Amount from Accounts where 'Auto_Reload' = 'true' and 'Hello_Hunter_Id' is not null ORDER BY Account_Name limit 200";
        const zCustomers = await this.crmApi.coql(query);
        logger.info('Initialized Zoho CRM Client');
        return zCustomers.data || [];
    }

    async processCustomers(zCustomers, hhCustomers) {
        const reloadDate = new Date().toISOString().split('T')[0];

        for (const zCust of zCustomers) {
            const zHhId = zCust.Hello_Hunter_Id.toString();
            const hhCust = hhCustomers.find(hhCust => hhCust.id.toString() === zHhId);

            if (!hhCust) {
                logger.error(`Hello Hunter customer not found for ID: ${zHhId}`);
                continue;
            }

            const balance = Math.round((hhCust.available_minutes * (hhCust.call_rate / 10000)) / 60);
            const custName = hhCust.user_group_name;
            const custThresh = zCust.Threshold;
            const reloadAmnt = zCust.Reload_Amount;

            logger.info(`Processing Customer: ${custName}, Balance: $${balance}, Threshold: $${custThresh}, Reload Amount: $${reloadAmnt}`);

            if (balance <= custThresh) {
                logger.info(`Balance below threshold. Reloading ${reloadAmnt} for ${custName}...`);
                const reloadResponse = await this.hhApi.giveCredit(zHhId, reloadAmnt);

                if (reloadResponse === 200) {
                    logger.info(`Successfully reloaded ${reloadAmnt} for ${custName}`);
                    await this.createReloadRecordCrm(zCust, reloadAmnt, balance, reloadDate, custName);
                } else {
                    logger.error(`Failed to reload ${reloadAmnt} for ${custName}`);
                }
            } else {
                logger.info(`${custName} is above the threshold. No reload required.`);
            }
        }
    }

    async createReloadRecordCrm(zCust, reloadAmnt, balance, reloadDate, custName) {
        const payload = [
            {
                Account: zCust.id,
                HH_Name: custName,
                New_Balance: balance + reloadAmnt,
                Old_Balance: balance,
                Platform_Id: zCust.Hello_Hunter_Id,
                Reload_Amount: reloadAmnt,
                Reload_Date: reloadDate,
            },
        ];

        logger.info('Creating reload record in Zoho CRM', payload);
        const reloadResponse = await this.crmApi.zohoCrmPost('Voice_Dial_Reloads', payload, 'upsert');

        if (reloadResponse !== 'fail') {
            logger.info(`CRM Response: ${reloadResponse[0]?.message || 'Success'}`);
        }
    }
}

module.exports = HelloHunterService;

File: Ruby/hh-credit-autoload-ruby/lambda_function.rb
require 'faraday'
require 'dotenv/load'
require_relative 'helpers/hello-hunter-api'
require_relative 'helpers/zoho-crm-api'

$logger = Logger.new($stdout)

def lambda_handler(event:, context:)
  z_custs = init_z_client
  hh_custs = init_hh_client
  process_customers(z_custs,hh_custs)
end

def init_hh_client
  $hh_api = HelloHunterApi.new
  hh_custs = $hh_api.customers
  hh_custs
end

def init_z_client
  $crm_api = ZohoCrmApi.new
  hh_query = "select Account_Name, Hello_Hunter_Id, Threshold, Reload_Amount from Accounts where 'Auto_Reload' = 'true' and 'Hello_Hunter_Id' is not null ORDER BY Account_Name limit 200"
  z_custs = $crm_api.coql(hh_query)['data']
  z_custs
end

def process_customers(z_custs,hh_custs)
  now_time = Time.new
  reload_date = now_time.strftime("%F")
  if z_custs != "fail"
    z_custs.each do |z_cust|
      $logger.info("Processing Customer: #{z_cust}")
      z_hh_id = z_cust['Hello_Hunter_Id'].to_s
      reload_cust = hh_custs.select{|hh_cust| hh_cust["id"].to_s.include?(z_hh_id)}[0]
      cust_call_rate = reload_cust['call_rate'].to_f / 10000
      cust_min = reload_cust['available_minutes'].to_f
      balance = (cust_min * cust_call_rate / 60).round(0)
      cust_name = reload_cust['user_group_name']
      cust_thresh = z_cust['Threshold']
        reload_amnt = z_cust['Reload_Amount']
      $logger.info("Customer: #{cust_name} Balance: $#{balance} Threshold $#{cust_thresh} Reload Amount: $#{reload_amnt}")

      if balance <= cust_thresh
        $logger.info("Below threshold. Need to reload.")
        hh_reload_resp = $hh_api.give_credit(z_hh_id,reload_amnt)
        if hh_reload_resp == 200
          $logger.info("Reloaded #{reload_amnt} to #{cust_name}")
          create_reload_record_crm(z_cust,reload_amnt,balance,reload_date,cust_name)
        else
          $logger.error("Failed to reload #{reload_amnt} to #{cust_name}")
        end
      else
        $logger.info("Above threshold. Don't need to reload")
      end
    end
  end
end

def create_reload_record_crm(z_cust,reload_amnt,balance,reload_date,cust_name)
  update_cust = {}
  update_cust['Account'] = z_cust['id']
  update_cust['HH_Name'] = cust_name
  update_cust['New_Balance'] = balance + reload_amnt
  update_cust['Old_Balance'] = balance
  update_cust['Platform_Id'] = z_cust['Hello_Hunter_Id']
  update_cust['Reload_Amount'] = reload_amnt
  update_cust['Reload_Date'] = reload_date

  payload = Array.new(1) {update_cust}
  $logger.info(payload)
  reload_resp = $crm_api.zoho_crm_post('Voice_Dial_Reloads', payload, '/upsert')
  $logger.info(reload_resp)
  if reload_resp != 'fail'
  $logger.info(reload_resp[0][0]['message'])
  end
end

def main
  lambda_handler(event: nil, context: nil)
end

main()
